### PR TITLE
Add package for simplified JWT usage

### PR DESCRIPTION
**Pull request by bruth**
This package will be shared between the auth service and proxies that must
authenticate requests before delegating to the downstream service or handler.

This package standardizes on using the RSA-256 signing method. The auth service
will create and sign tokens using the private key and any consumer which requires
authenticating a token will have a public key used to verify the token.

Note that JWTs trade-off revocation for decentralization. In the use for Arcus
revocation will be an uncommon need, however request verification must occur
all the time. The decentralization property allows for local verification using
the public key without requiring a network call.

---

For more context, the primary use case for JWTs will be for Lab deployments. There is an Envoy proxy that binds DNS entries to upstream servers (e.g. scit12-sepsis.lab.arcus.chop.edu -> the deployed pod). This proxy will delegate authentication and authorization of requests to a sidecar service (Envoy has a native extauthz API for this). This sidecar service will verify the JWT from the Authorization or Cookie header as well as check if the authenticated user can access the lab they are requesting. This ACL will be defined as an OPA policy.

### CODE REVIEW

**Comment by bruth** [comment]
Thanks @dmead 

> is it wise to have keys in a test? these guys probably have expiration dates (i'm assuming) which mean the tests will eventually fail at some point in the future it may be safer to generate keys at test time and throw them away every run.

The keys won't expire, but that is a good point to generate them using the native Go libraries two less artifacts in the repo.

> it would be good to put this in an interface that can be mixed in with other service types. go is statically typed so lets take advantage of that.

I am not quite sure what you mean. Do you mean create a Go interface with `Sign` and `Parse` defined? What do you mean by service type?

**Comment by bruth** [comment]
Addressed the in-memory key pair generation.

**Comment by dmead** [comment]
> Thanks @dmead
> 
> > is it wise to have keys in a test? these guys probably have expiration dates (i'm assuming) which mean the tests will eventually fail at some point in the future it may be safer to generate keys at test time and throw them away every run.
> 
> The keys won't expire, but that is a good point to generate them using the native Go libraries for two less artifacts in the repo.
> 
> > it would be good to put this in an interface that can be mixed in with other service types. go is statically typed so lets take advantage of that.
> 
> I am not quite sure what you mean. Do you mean create a Go interface with `Sign` and `Parse` defined? What do you mean by service type?


mixed was the wrong word. I should have said composed or embedded. Assuming the auth service and proxies you mentioned in the commit message have interfaces, this interface for parsing and signing can embedded in part of those to reflect this dependency in the type system. like so

https://www.ardanlabs.com/blog/2015/09/composition-with-go.html

File: jwt/jwt.go [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+	Audience  string `json:"aud,omitempty"`
+	ExpiresAt int64  `json:"exp,omitempty"`
+	Id        string `json:"jti,omitempty"`
+	IssuedAt  int64  `json:"iat,omitempty"`
+	Issuer    string `json:"iss,omitempty"`
+	NotBefore int64  `json:"nbf,omitempty"`
+	Subject   string `json:"sub,omitempty"`
+}
+
+// Sign creates and signs a JWT given the claims. It uses the HS256 signing method
```
Comment (murphyke) [inline jwt/jwt.go]: The comment says `HS256`, but the code uses `RS256`.

File: jwt/jwt.go [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+func ReadPublicKeyPEM(path string) (*rsa.PublicKey, error) {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	return ParsePublicKeyPEM(b)
+}
+
+type Claims struct {
```
Comment (murphyke) [inline jwt/jwt.go]: What's the difference between `jwt.StandardClaims` and this `Claims`? The code is using both ....

Comment (bruth) [reply jwt/jwt.go]: Heh.. good question. The `jwt.StandardClaims` is a struct type that implements `jwt.Claims` (an interface). As part of this package I didn't want any of the `jwt` package to leak as a dependency into the caller package. So simply created a local corresponding type that gets converted to the correspond `jwt` claim type (there are a few). Also going forward, it is possible to add more claims that could be Arcus-specific. So its a touch of reducing caller dependencies and future-proofing.

Comment (murphyke) [reply jwt/jwt.go]: > [StandardClaims/Claims]

@bruth Makes perfect sense to me, thanks.

File: jwt/jwt.go (hunk 1 above)
Comment (bruth) [reply jwt/jwt.go]: Fixed this.

Comment (murphyke) [reply jwt/jwt.go]: So, having just read your initial PR comment, I think the `HS256` in the comment must be a typo.

**Comment by bruth** [comment]
> this interface for parsing and signing can embedded in those to reflect this dependency in the type system

@dmead Thanks for the link. I am aware of composition, but I don't think it is necessary in this case. Specifically, defining an interface and embedding it into the struct modeling the service wouldn't really add much benefit to this use case.

Just to reiterate the workflow, the auth service would handle producing and signing the JWT (using the function in this PR) when the OAuth callback occurs (on successful authentication by the user). The response would be a redirect back to the original endpoint along with a `set-cookie` header (since all of this is done in the browser). The browser would receive this response, set the cookie and then perform a request to the original location. At this point the proxy receives the request again and needs to authorize it. This time it will see the cookie and will verify it (using the function in this PR) followed by doing the OPA access control check. Assuming both are ok, the request will be allowed.

File: jwt/jwt.go (hunk 1 above)
Comment (bruth) [reply jwt/jwt.go]: Sorry, yes RS256 is the correct one. I did fix the comment.

**Comment by bruth** [comment]
I believe I addressed all the comments. Any more feedback?

**Comment by murphyke** [comment]
Not from me. I'll let @dmead approve or not.

**Comment by bruth** [comment]
@dmead Can you approve if you have no other comments?

**Comment by bruth** [comment]
Merging to move forward.

**Pull request merged**